### PR TITLE
Vite: Replace vite-plugin-externals with custom plugin

### PR DESCRIPTION
**Pull request by Dschungelabenteuer**
Issue: #20643

## What I did

* When used in dev mode, `vite-plugin-externals` breaks HMR for vite-based projects.
* When disabled in dev mode, it completely prevents Storybook from working on vite + pnpm (and probably Yarn PnP) projects.
* I replaced that plugin with a local one, which is straightforward and quite naive

## What I'm not sure I checked correctly

* Production build works fine
* That the #20514 issue about code duplication is still fixed (I've searched for 

## What I'm sure I didn't checked
* Yarn PnP

## How to test

- You can create any Vite sandbox, but it will still require you to change the package manager to pnpm so that dependencies do not get hoisted.
- Make sure you get the following error: `[vite] Internal server error: Failed to resolve import "@storybook/preview-api" from "../../../../../../virtual:/@storybook/builder-vite/vite-app.js". `
- Build the `@storybook/builder-vite` package from the source branch of this PR
- Copy the generated `dist` output of `@storybook/builder-vite` to replace the one in your pnpm workspace 
- Start storybook
- Be happy because the error is gone!

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Comment by IanVS** [comment]
This is working fine for me locally, but is failing in CI.  @ndelangen maybe you can take a look?  I've confirmed this approach does avoid duplicate imports, and HMR works correctly.  But I don't know why the build fails in CI and not with the same command locally.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * ```
+ *
+ * becomes
+ *
+ * ```js
+ * const { useMemo: useMemo2, useEffect: useEffect2 } = __STORYBOOK_MODULE_PREVIEW_API__;
+ * ```
+ *
+ * It is based on existing plugins like https://github.com/crcong/vite-plugin-externals
+ * and https://github.com/eight04/rollup-plugin-external-globals, but simplified to meet our meager needs.
```
Comment (ndelangen) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:28]:
```suggestion
 * and https://github.com/eight04/rollup-plugin-external-globals, but simplified to meet our manager needs.
```

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      await ensureDir(cachePath);
+      await emptyDirSync(cachePath);
+
+      // eslint-disable-next-line no-restricted-syntax
+      for await (const externalKey of Object.keys(globals) as Array<keyof typeof globals>) {
+        const externalCachePath = join(cachePath, `${externalKey}.js`);
+        newAlias.push({ find: new RegExp(`^${externalKey}$`), replacement: externalCachePath });
+        await ensureFile(externalCachePath);
+        await writeFile(externalCachePath, `module.exports = ${globals[externalKey]};`);
+      }
```
Comment (ndelangen) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:52]: This runs each in series, it awaits at every operation.

```
ensure 'a'
write 'a'
ensure 'b'
write 'b'
... etc
```

I think we can do better with a `Promise.all` running all this in parallel.
If this list of files gets larger, performance will suffer.

Key to performance here is to do as much work in parallel as possible.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+              .slice(startPosition, endPosition)
+              .replace('import ', 'const ')
+              .replace('import{', 'const {')
+              .replaceAll(' as ', ': ')
+              .replace(' from ', ' = ')
+              .replace('}from', '} = ')
+              .replace(packageAndDelimiters, globals[packageName])
+          );
+        }
+      });
```
Comment (ndelangen) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:84]: I am scared, without a proper test-suite. ... we have tests for this?

We should extract this code into a function and test the flip out of it.

I would also recommend writing a single regex + a single replace, that's easier to grok than 6 separate replace commands.?

**Review by ndelangen — COMMENTED** [review]
Needs test, needs perf improvements.

I can work on this if needed.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts (hunk 1 above)
Comment (IanVS) [reply code/lib/builder-vite/src/plugins/externals-plugin.ts:28]: I did mean "meager" here, as in small and not very demanding. We could replace it with "simple".

**Comment by IanVS** [comment]
Thanks, I completely agree. Just wanted to prove out that the approach would work first, and it seems like it will. If you have time to work on it that would be great, otherwise I'll try to take a shot at it tonight. Or maybe @Dschungelabenteuer would like to?

**Comment by Dschungelabenteuer** [comment]
@IanVS I can try while you take some rest until you feel better :) Eventually I'll probably need some guidance around tests, though!

**Comment by stevezhu** [comment]
I can also help on anything here if needed.

**Comment by stevezhu** [comment]
Don't mean to derail this implementation, but it's also possible to instead more simply wrap the original plugin within a custom plugin, allowing to fix bugs/swap out certain parts or extend.

```typescript
import type { Plugin } from 'vite';
import { globals } from '@storybook/preview/globals';

import { viteExternalsPlugin } from 'vite-plugin-externals';

export async function externalsPlugin() {
  const plugin = viteExternalsPlugin(globals, { useWindow: false });
  const config = plugin.config as Extract<
    ReturnType<typeof viteExternalsPlugin>['config'],
    (...args: any[]) => any
  >;
  return {
    ...plugin,
    name: 'storybook:externals-plugin',
    async config(...configArgs) {
      const conf = await config(...configArgs);
      return {
        resolve: {
          alias: conf?.resolve?.alias,
        },
      };
    },
  } satisfies Plugin;
}
```

**Comment by ndelangen** [comment]
@Dschungelabenteuer Do you want @stevezhu to take over?

I think, with a bit of cleanup this PR is good to go.
And this has high priority for me to get merged.

**Comment by Dschungelabenteuer** [comment]
> @Dschungelabenteuer Do you want @stevezhu to take over?
> 
> I think, with a bit of cleanup this PR is good to go. And this has high priority for me to get merged.

I'm still at work so if @stevezhu wants to give it a shot, please feel free :)

**Comment by ndelangen** [comment]
High priority means, "preferably this week" 😄

**Comment by stevezhu** [comment]
PR for the suggestion here: https://github.com/storybookjs/storybook/pull/20770

Simple performance test @IanVS in one of the comments https://github.com/storybookjs/storybook/pull/20770#issuecomment-1402221289

**Comment by IanVS** [comment]
Thanks @stevezhu.  I wanted to get a bit more granular perf numbers for just the plugin, so I pushed up two branches, https://github.com/storybookjs/storybook/tree/timing/external-vite-plugin and https://github.com/storybookjs/storybook/tree/timing/internal-vite-plugin.  I added timings to the `config` and `transform` hooks separately, and this is roughly what I found in the react-ts sandbox:

external (wrapped dependency):
- config: 0.05 sec
- transform: 0.65 sec

internal:
- config: 4 sec
- transform: 0.02 sec

This surprised me, since the config hooks are nearly identical between the two.  But, the external plugin was using a newer version of `fs-extra`, and when I updated storybook, the times matched up.  So I submitted https://github.com/storybookjs/storybook/pull/20772 because I think we'll want to update fs-extra no matter what.


At any rate, it does seem like the approach in this PR is slightly more performant, and we should keep in mind that the sandbox is a small project, and we could expect the transform time to grow with the number of files being transformed.  There's also something nice about controlling our own destiny and not relying on another maintainer to merge fixes, though the workaround in https://github.com/storybookjs/storybook/pull/20770 is clever and would indeed do what we need.  

In summary, I think either approach would be fine, the internal plugin is a bit faster (after the fs-extra upgrade), but if we don't feel comfortable with its robustness we could use the wrapped dependency instead.

**Comment by stevezhu** [comment]
The fs-extra performance was a great find. Ultimately up to you on what you think will be the better solution; just wanted to offer up the other option. Seeing as how the external plugin is relatively unmaintained and the performance gains I'm definitely leaning towards having an internal plugin.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+type SingleGlobalName = keyof typeof globals;
+type Globals = typeof globals & Record<string, string>;
+
+const replacementMap = new Map([
+  ['import ', 'const '],
+  ['import{', 'const {'],
+  [' as ', ': '],
+  [' from ', ' = '],
+  ['}from', '} ='],
+]);
```
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:18]: Does commonjs require need to be handled as well?

**Comment by Dschungelabenteuer** [comment]
Nice insight from both of you guys, I was quite curious about the transform hook which seemed like the most relevant concern, especially because it looks like candidate files are being traversed twice (first by `es-module-lexer` just like in the current PR, then with `acorn` to replace tokens based on file's AST, which is safer but also heavier. I would never have suspected potential performance issues inside the config hook, nice catch!

Do we have a clear idea of the amount of files served by Vite that are actually transformed by the plugin? I mean, there are the obvious virtual files such as `vite-app.js` and `setup-addons.js`, but is there something else we should be paying close attention to? Maybe this could help consolidate our transform process so that we feel more confident about this?

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        }),
+      };
+    },
+  } satisfies Plugin;
+}
+
+export function rewriteImport<T = true>(
+  importStatement: string,
+  globs: T extends true ? Globals : Record<string, string>,
+  packageName: keyof typeof globs & string
```
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:100]: It doesn't seem like there's any value in this generic. Seems like it can just be this instead:
```suggestion
export function rewriteImport<T = true>(
  importStatement: string,
  globs: Record<string, string>,
  packageName: string
```

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    },
+    // Replace imports with variables destructured from global scope
+    async transform(code: string, id: string) {
+      const globalsList = Object.keys(globals) as SingleGlobalName[];
+      if (globalsList.every((glob) => !code.includes(glob))) return undefined;
+
+      const [imports] = parse(code);
+      const src = new MagicString(code);
+      imports.forEach(({ n: path, ss: startPosition, se: endPosition }) => {
+        const packageName = path as SingleGlobalName | undefined;
```
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:77]: No point in adding these casts I think. I doesn't increase type safety.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+      return {
+        resolve: {
+          alias: newAlias,
+        },
+      };
+    },
+    // Replace imports with variables destructured from global scope
+    async transform(code: string, id: string) {
+      const globalsList = Object.keys(globals) as SingleGlobalName[];
```
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:71]: I don't think there's any point in adding this cast. It doesn't really increase type safety.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+ * becomes
+ *
+ * ```js
+ * const { useMemo: useMemo2, useEffect: useEffect2 } = __STORYBOOK_MODULE_PREVIEW_API__;
+ * ```
+ *
+ * It is based on existing plugins like https://github.com/crcong/vite-plugin-externals
+ * and https://github.com/eight04/rollup-plugin-external-globals, but simplified to meet our simple needs.
+ */
+export async function externalsPlugin() {
```
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:39]: Personally, even though this is only used for this one case right now, globals should probably still be an argument for this plugin instead of being imported directly from the `@storybook/preview/globals` package. More flexible for reuse in the case that the args need to be changed or this plugin needs to be used for another set of globals.

Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:39]: Personally, even though this is only used for this one case right now, globals should probably still be an argument for this plugin instead of being imported directly from the `@storybook/preview/globals` package. More flexible for reuse in the case that the args need to be changed or this plugin needs to be used for another set of globals.
```suggestion
export async function externalsPlugin(externals: Record<string, string>) {
```

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  await init;
+  return {
+    name: 'storybook:externals-plugin',
```
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:42]: This is a small thing, but do you think this would be clearer named as `external-globals-plugin` similar to the rollup plugin? Typically externals for the build process doesn't include transforming imports to use global variables.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts (hunk 8 above)
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:39]: Personally, even though this is only used for this one case right now, globals should probably still be an argument for this plugin instead of being imported directly from the `@storybook/preview/globals` package. More flexible for reuse in the case that the args need to be changed or this plugin needs to be used for another set of globals.
```suggestion
export async function externalsPlugin(externals: Record<string, string>) {
```

File: code/lib/builder-vite/src/plugins/externals-plugin.ts (hunk 9 above)
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:42]: This is a small thing, but do you think this would be clearer named as `external-globals-plugin` similar to the rollup plugin? Since typically externals for the build process doesn't include transforming imports to use global variables.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts (hunk 7 above)
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:71]: I don't think there's any point in adding this cast. It doesn't really increase type safety.

File: code/lib/builder-vite/src/plugins/externals-plugin.ts (hunk 6 above)
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:77]: same here

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      };
+    },
+  } satisfies Plugin;
+}
+
+export function rewriteImport<T = true>(
+  importStatement: string,
+  globs: T extends true ? Globals : Record<string, string>,
+  packageName: keyof typeof globs & string
+): string {
```
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:101]: It doesn't seem like there's any value in this generic right? Seems like it can just be this instead:
```suggestion
export function rewriteImport(
  importStatement: string,
  globs: Record<string, string>,
  packageName: string
): string {
```

**Review by stevezhu — COMMENTED** [review]
A few small comments

File: code/lib/builder-vite/src/plugins/externals-plugin.ts (hunk 9 above)
Comment (stevezhu) [reply code/lib/builder-vite/src/plugins/externals-plugin.ts]: A yep that's what I meant, I omitted writing out `storybook:` portion.

File: code/lib/builder-vite/src/plugins/externals-plugin.test.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  {
+    globals,
+    packageName,
+    input: `import { Rain, Jour as Day, Nuit as Night, Sun } from "${packageName}"`,
+    output: `const { Rain, Jour: Day, Nuit: Night, Sun } = ${globals[packageName]}`,
+  },
+  {
+    globals,
+    packageName,
+    input: `import{Rain,Jour as Day,Nuit as Night,Sun}from '${packageName}'`,
```
Comment (IanVS) [inline code/lib/builder-vite/src/plugins/externals-plugin.test.ts:16]: The space after `from` can be deleted, since there's a quotation mark.

File: code/lib/builder-vite/src/plugins/externals-plugin.test.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    output: `const {Rain,Jour: Day,Nuit: Night,Sun} = ${globals[packageName]}`,
+  },
+  {
+    globals,
+    packageName,
+    input: `const { Afternoon } = await import('${packageName}')`,
+    output: `const { Afternoon } = ${globals[packageName]}`,
+  },
```
Comment (IanVS) [inline code/lib/builder-vite/src/plugins/externals-plugin.test.ts:24]: We should also handle `import * as foo from "package"`.  We might not have any of these right now, but we could encounter it in the future and we wouldn't want this to break.  It's maybe a bit trickier to handle, but we'd want it to become `const foo = __STORYBOOK_MODULE_PREVIEW_API__`.

File: code/lib/builder-vite/src/plugins/externals-plugin.test.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+];
```
Comment (IanVS) [inline code/lib/builder-vite/src/plugins/externals-plugin.test.ts:25]: We need a test for default imports, and I think we might be handling them incorrectly right now.  Correct me if I'm wrong, but I think we should convert from 

```js
import foo from '@storybook/preview-api';
```
to 
```js
const {default: foo} = __STORYBOOK_MODULE_PREVIEW_API__;
```

File: code/lib/builder-vite/src/plugins/externals-plugin.ts (hunk 4 above)
Comment (IanVS) [reply code/lib/builder-vite/src/plugins/externals-plugin.ts:17]: I don't believe so.   None of the code processed by Vite will be using CJS.

**Review by IanVS — CHANGES_REQUESTED** [review]
I think this is really close, but there are a couple of cases we should handle (and test).

File: code/lib/builder-vite/src/plugins/externals-plugin.ts (hunk 8 above)
Comment (ndelangen) [reply code/lib/builder-vite/src/plugins/externals-plugin.ts:41]: Yes, I agree, that is just nice separation of concerns!

**Review by IanVS — APPROVED** [review]
This looks good to me, I tested it out in a new sandbox and `yarn storybook` starts fine and HMR still works.  I'd love if we can get this merged and released ASAP since it fixes a pretty large issue for vite (and especially pnpm) users.  Thanks for all the great work here @Dschungelabenteuer and for the help/reviews/ideas @stevezhu.  🙌 🚀 

@ndelangen how's it look to you?

File: code/lib/builder-vite/src/plugins/externals-plugin.ts [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+ * becomes
+ *
+ * ```js
+ * const { useMemo: useMemo2, useEffect: useEffect2 } = __STORYBOOK_MODULE_PREVIEW_API__;
+ * ```
+ *
+ * It is based on existing plugins like https://github.com/crcong/vite-plugin-externals
+ * and https://github.com/eight04/rollup-plugin-external-globals, but simplified to meet our simple needs.
+ */
+export async function externalsPlugin(externals: Record<string, string>) {
```
Comment (stevezhu) [inline code/lib/builder-vite/src/plugins/externals-plugin.ts:38]: Should probably change this as well along with the filename for uniformity `external-globals-plugin.ts`
```suggestion
export async function externalGlobalsPlugin(externals: Record<string, string>) {
```

Comment (IanVS) [reply code/lib/builder-vite/src/plugins/externals-plugin.ts:38]: We can also do this in a followup if we just want to get it merged quickly.  I wouldn't treat it as a blocker, personally.

Comment (stevezhu) [reply code/lib/builder-vite/src/plugins/externals-plugin.ts:38]: Agreed

**Comment by dryton** [comment]
I'm seeing an issue in relation to this change in beta.36:
```
11:03:43 AM [vite] Internal server error: src.update is not a function
  Plugin: storybook:external-globals-plugin
  File: /virtual:/@storybook/builder-vite/vite-app.js
      at null.<anonymous> (/Users/name/project/node_modules/@storybook/builder-vite/dist/index.js:162:4407)
      at Array.forEach (<anonymous>)
      at TransformContext.transform (/Users/name/project/node_modules/@storybook/builder-vite/dist/index.js:162:4148)
      at Object.transform (file:///Users/name/project/node_modules/vite/dist/node/chunks/dep-665b0112.js:41031:44)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async loadAndTransform (file:///Users/name/project/node_modules/vite/dist/node/chunks/dep-665b0112.js:37292:29)
11:03:45 AM [vite] error while updating dependencies:
Error: ENOENT: no such file or directory, rename '/Users/name/project/node_modules/.cache/.vite-storybook/deps_temp' -> '/Users/name/project/node_modules/.cache/.vite-storybook/deps'
    at renameSync (node:fs:980:3)
    at Object.commit (file:///Users/name/project/node_modules/vite/dist/node/chunks/dep-665b0112.js:42219:19)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async commitProcessing (file:///Users/name/project/node_modules/vite/dist/node/chunks/dep-665b0112.js:41694:17)
    at async runOptimizer (file:///Users/name/project/node_modules/vite/dist/node/chunks/dep-665b0112.js:41732:17)
```

**Comment by IanVS** [comment]
@dryton it looks like you might be hitting https://github.com/vitejs/vite/issues/9986 or some variation on it. Are you able to reproduce it reliably, or does it only happen occasionally?  If you can reproduce it, can you please open an issue with a link to a reproduction?

**Comment by Dschungelabenteuer** [comment]
@dryton as mentioned in [this discussion](https://discord.com/channels/486522875931656193/1069653963521150986/1069666510739619840), I think we'll have to upgrade the minimum expected `magic-string` version, since the `update` method only exists since `0.26.6` and the `@storybook/builder-vite` only sets it to `^0.26.1`.
